File: utils/data_processor.py
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
from typing import List, Dict, Any
import streamlit as st

class DataProcessor:
    """Process Stripe subscription data to calculate MRR metrics"""
    
    def __init__(self):
        """Initialize data processor"""
        pass
    
    def calculate_new_mrr_by_month(self, subscriptions: List[Dict[Any, Any]]) -> pd.DataFrame:
        """
        Calculate New Monthly Recurring Revenue by month from subscription data
        
        Args:
            subscriptions: List of Stripe subscription objects
            
        Returns:
            DataFrame with columns: month, new_mrr, subscription_count
        """
        if not subscriptions:
            return pd.DataFrame()
        
        try:
            # Extract relevant subscription data
            subscription_data = []
            
            for sub in subscriptions:
                # Get subscription start date
                created_date = datetime.fromtimestamp(sub.get('created', 0))
                
                # Calculate MRR for this subscription
                mrr_amount = self._calculate_subscription_mrr(sub)
                
                if mrr_amount > 0:  # Only include subscriptions with positive MRR
                    subscription_data.append({
                        'subscription_id': sub.get('id'),
                        'created_date': created_date,
                        'mrr_amount': mrr_amount,
                        'status': sub.get('status'),
                        'month_year': created_date.strftime('%Y-%m')
                    })
            
            if not subscription_data:
                return pd.DataFrame()
            
            # Create DataFrame
            df = pd.DataFrame(subscription_data)
            
            # Group by month and calculate new MRR
            monthly_mrr = df.groupby('month_year').agg({
                'mrr_amount': 'sum',
                'subscription_id': 'count'
            }).reset_index()
            
            # Rename columns
            monthly_mrr.columns = ['month_year', 'new_mrr', 'subscription_count']
            
            # Convert month_year to proper date format for sorting
            monthly_mrr['month_date'] = pd.to_datetime(monthly_mrr['month_year'])
            monthly_mrr = monthly_mrr.sort_values('month_date')
            
            # Format month for display
            monthly_mrr['month'] = monthly_mrr['month_date'].dt.strftime('%b %Y')
            
            # Select final columns
            result_df = monthly_mrr[['month', 'new_mrr', 'subscription_count']].copy()
            
            return result_df
            
        except Exception as e:
            st.error(f"Error processing subscription data: {str(e)}")
            return pd.DataFrame()
    
    def _calculate_subscription_mrr(self, subscription: Dict[Any, Any]) -> float:
        """
        Calculate Monthly Recurring Revenue for a single subscription
        
        Args:
            subscription: Stripe subscription object
            
        Returns:
            Monthly recurring revenue amount
        """
        try:
            total_mrr = 0
            
            # Get subscription items
            items = subscription.get('items', {}).get('data', [])
            
            for item in items:
                price = item.get('price', {})
                quantity = item.get('quantity', 1)
                
                if not price:
                    continue
                
                # Get price amount and interval
                unit_amount = price.get('unit_amount', 0) / 100  # Convert from cents
                interval = price.get('recurring', {}).get('interval')
                interval_count = price.get('recurring', {}).get('interval_count', 1)
                
                # Convert to monthly amount
                if interval == 'month':
                    monthly_amount = unit_amount / interval_count
                elif interval == 'year':
                    monthly_amount = unit_amount / (12 * interval_count)
                elif interval == 'week':
                    monthly_amount = unit_amount * (52 / 12) / interval_count
                elif interval == 'day':
                    monthly_amount = unit_amount * (365 / 12) / interval_count
                else:
                    # Unknown interval, skip
                    continue
                
                total_mrr += monthly_amount * quantity
            
            return round(total_mrr, 2)
            
        except Exception as e:
            # Log error but don't fail the entire calculation
            print(f"Error calculating MRR for subscription {subscription.get('id', 'unknown')}: {str(e)}")
            return 0
    
    def calculate_mrr_growth(self, df: pd.DataFrame) -> Dict[str, float]:
        """
        Calculate MRR growth metrics
        
        Args:
            df: DataFrame with MRR data by month
            
        Returns:
            Dictionary with growth metrics
        """
        if df.empty or len(df) < 2:
            return {
                'current_month_mrr': 0,
                'previous_month_mrr': 0,
                'month_over_month_growth': 0,
                'total_new_mrr': 0
            }
        
        try:
            current_month_mrr = df['new_mrr'].iloc[-1]
            previous_month_mrr = df['new_mrr'].iloc[-2]
            total_new_mrr = df['new_mrr'].sum()
            
            # Calculate growth rate
            if previous_month_mrr > 0:
                growth_rate = ((current_month_mrr - previous_month_mrr) / previous_month_mrr) * 100
            else:
                growth_rate = 100 if current_month_mrr > 0 else 0
            
            return {
                'current_month_mrr': current_month_mrr,
                'previous_month_mrr': previous_month_mrr,
                'month_over_month_growth': round(growth_rate, 2),
                'total_new_mrr': total_new_mrr
            }
            
        except Exception as e:
            st.error(f"Error calculating growth metrics: {str(e)}")
            return {
                'current_month_mrr': 0,
                'previous_month_mrr': 0,
                'month_over_month_growth': 0,
                'total_new_mrr': 0
            }
    
    def calculate_arr(self, subscriptions: List[Dict[Any, Any]]) -> float:
        """
        Calculate Annual Recurring Revenue (ARR)
        
        Args:
            subscriptions: List of Stripe subscription objects
            
        Returns:
            Total ARR
        """
        total_arr = 0
        
        for sub in subscriptions:
            if sub.get('status') in ['active', 'trialing']:
                mrr = self._calculate_subscription_mrr(sub)
                total_arr += mrr * 12
        
        return round(total_arr, 2)
    
    def calculate_churn_metrics(self, subscriptions: List[Dict[Any, Any]]) -> Dict[str, Any]:
        """
        Calculate churn rate and related metrics
        
        Args:
            subscriptions: List of Stripe subscription objects
            
        Returns:
            Dictionary with churn metrics (churn for current month period)
        """
        try:
            from datetime import datetime
            from dateutil.relativedelta import relativedelta
            
            current_date = datetime.now()
            start_of_current_month = current_date.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
            
            active_start_of_month = 0
            active_current = 0
            churned_this_month = 0
            
            for sub in subscriptions:
                created_date = datetime.fromtimestamp(sub.get('created', 0))
                canceled_date = None
                
                if sub.get('canceled_at'):
                    canceled_date = datetime.fromtimestamp(sub.get('canceled_at'))
                
                was_active_at_month_start = (
                    created_date < start_of_current_month and
                    (not canceled_date or canceled_date >= start_of_current_month)
                )
                
                is_active_now = (
                    created_date < current_date and
                    (not canceled_date or canceled_date >= current_date)
                )
                
                if was_active_at_month_start:
                    active_start_of_month += 1
                    
                    if not is_active_now and canceled_date and start_of_current_month <= canceled_date < current_date:
                        churned_this_month += 1
                
                if is_active_now:
                    active_current += 1
            
            churn_rate = (churned_this_month / active_start_of_month * 100) if active_start_of_month > 0 else 0
            
            return {
                'churn_rate': round(churn_rate, 2),
                'churned_count': churned_this_month,
                'active_count': active_current,
                'previous_active_count': active_start_of_month
            }
            
        except Exception as e:
            return {
                'churn_rate': 0,
                'churned_count': 0,
                'active_count': 0,
                'previous_active_count': 0
            }
    
    def calculate_customer_trends(self, subscriptions: List[Dict[Any, Any]]) -> pd.DataFrame:
        """
        Calculate customer count trends by month
        
        Args:
            subscriptions: List of Stripe subscription objects
            
        Returns:
            DataFrame with customer counts by month
        """
        try:
            customer_data = []
            
            for sub in subscriptions:
                created_date = datetime.fromtimestamp(sub.get('created', 0))
                month_year = created_date.strftime('%Y-%m')
                
                customer_data.append({
                    'month_year': month_year,
                    'customer_id': sub.get('customer'),
                    'created_date': created_date
                })
            
            if not customer_data:
                return pd.DataFrame()
            
            df = pd.DataFrame(customer_data)
            
            monthly_customers = df.groupby('month_year').agg({
                'customer_id': 'count'
            }).reset_index()
            
            monthly_customers.columns = ['month_year', 'new_customers']
            
            monthly_customers['month_date'] = pd.to_datetime(monthly_customers['month_year'])
            monthly_customers = monthly_customers.sort_values('month_date')
            
            monthly_customers['month'] = monthly_customers['month_date'].dt.strftime('%b %Y')
            monthly_customers['cumulative_customers'] = monthly_customers['new_customers'].cumsum()
            
            return monthly_customers[['month', 'new_customers', 'cumulative_customers']]
            
        except Exception as e:
            st.error(f"Error calculating customer trends: {str(e)}")
            return pd.DataFrame()
    
    def calculate_new_vs_existing_mrr_by_month(self, subscriptions: List[Dict[Any, Any]]) -> pd.DataFrame:
        """
        Calculate New MRR vs Existing MRR by month
        
        Args:
            subscriptions: List of Stripe subscription objects
            
        Returns:
            DataFrame with columns: month, new_mrr, existing_mrr, total_mrr
        """
        if not subscriptions:
            return pd.DataFrame()
        
        try:
            from dateutil.relativedelta import relativedelta
            
            subscription_data = []
            
            for sub in subscriptions:
                created_date = datetime.fromtimestamp(sub.get('created', 0))
                canceled_at = sub.get('canceled_at')
                canceled_date = datetime.fromtimestamp(canceled_at) if canceled_at else None
                status = sub.get('status')
                mrr_amount = self._calculate_subscription_mrr(sub)
                
                if mrr_amount > 0:
                    subscription_data.append({
                        'subscription_id': sub.get('id'),
                        'created_date': created_date,
                        'canceled_date': canceled_date,
                        'status': status,
                        'mrr_amount': mrr_amount
                    })
            
            if not subscription_data:
                return pd.DataFrame()
            
            df = pd.DataFrame(subscription_data)
            
            all_months = set()
            for _, row in df.iterrows():
                created_month = row['created_date'].replace(day=1, hour=0, minute=0, second=0, microsecond=0)
                all_months.add(created_month)
                
                if row['canceled_date']:
                    canceled_month = row['canceled_date'].replace(day=1, hour=0, minute=0, second=0, microsecond=0)
                    current = created_month
                    while current <= canceled_month:
                        all_months.add(current)
                        current = current + relativedelta(months=1)
                else:
                    current = created_month
                    now = datetime.now().replace(day=1, hour=0, minute=0, second=0, microsecond=0)
                    while current <= now:
                        all_months.add(current)
                        current = current + relativedelta(months=1)
            
            monthly_results = []
            
            for month in sorted(all_months):
                month_end = month + relativedelta(months=1) - timedelta(seconds=1)
                new_mrr = 0
                existing_mrr = 0
                
                for _, row in df.iterrows():
                    created_date = row['created_date']
                    canceled_date = row['canceled_date']
                    mrr = row['mrr_amount']
                    
                    is_active_in_month = (
                        created_date <= month_end and
                        (not canceled_date or canceled_date >= month)
                    )
                    
                    if is_active_in_month:
                        created_month = created_date.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
                        if created_month == month:
                            new_mrr += mrr
                        else:
                            existing_mrr += mrr
                
                monthly_results.append({
                    'month_date': month,
                    'new_mrr': round(new_mrr, 2),
                    'existing_mrr': round(existing_mrr, 2),
                    'total_mrr': round(new_mrr + existing_mrr, 2)
                })
            
            result_df = pd.DataFrame(monthly_results)
            result_df = result_df.sort_values('month_date')
            result_df['month'] = result_df['month_date'].dt.strftime('%b %Y')
            
            return result_df[['month', 'new_mrr', 'existing_mrr', 'total_mrr']].copy()
            
        except Exception as e:
            st.error(f"Error calculating new vs existing MRR: {str(e)}")
            return pd.DataFrame()
    
    def calculate_revenue_by_plan(self, subscriptions: List[Dict[Any, Any]], product_names: Dict[str, str] = None) -> pd.DataFrame:
        """
        Calculate revenue breakdown by plan/product
        
        Args:
            subscriptions: List of Stripe subscription objects
            product_names: Optional dictionary mapping product IDs to product names
            
        Returns:
            DataFrame with revenue by plan
        """
        try:
            if product_names is None:
                product_names = {}
            
            plan_data = []
            
            for sub in subscriptions:
                if sub.get('status') not in ['active', 'trialing']:
                    continue
                
                items = sub.get('items', {}).get('data', [])
                
                for item in items:
                    price = item.get('price', {})
                    product = price.get('product', {})
                    
                    # Determine product name
                    if isinstance(product, dict):
                        # Product was expanded, use the name from it
                        product_name = product.get('name', 'Unknown Product')
                    elif isinstance(product, str):
                        # Product is an ID, look up the name from product_names map
                        product_name = product_names.get(product, product)
                    else:
                        product_name = 'Unknown Product'
                    
                    mrr = self._calculate_subscription_mrr({'items': {'data': [item]}})
                    
                    plan_data.append({
                        'product_name': product_name,
                        'mrr': mrr
                    })
            
            if not plan_data:
                return pd.DataFrame()
            
            df = pd.DataFrame(plan_data)
            
            plan_summary = df.groupby('product_name').agg({
                'mrr': 'sum'
            }).reset_index()
            
            plan_summary.columns = ['plan', 'mrr']
            plan_summary = plan_summary.sort_values('mrr', ascending=False)
            plan_summary['percentage'] = (plan_summary['mrr'] / plan_summary['mrr'].sum() * 100).round(2)
            
            return plan_summary
            
        except Exception as e:
            st.error(f"Error calculating revenue by plan: {str(e)}")
            return pd.DataFrame()

File: nextjs-dashboard/lib/data-processor.ts
import Stripe from 'stripe';
import { format, startOfMonth, endOfMonth, addMonths } from 'date-fns';

export interface MonthlyMRR {
  month: string;
  monthDate: Date;
  newMRR: number;
  existingMRR: number;
  totalMRR: number;
}

export interface CustomerTrend {
  month: string;
  monthDate: Date;
  newCustomers: number;
  cumulativeCustomers: number;
}

export interface PlanRevenue {
  plan: string;
  mrr: number;
  percentage: number;
}

export interface ChurnMetrics {
  churnRate: number;
  churnedCount: number;
  activeCount: number;
  previousActiveCount: number;
}

export interface FinancialMetrics {
  grossRevenue: number;
  stripeFees: number;
  netRevenue: number;
  totalPayouts: number;
  pendingBalance: number;
  availableBalance: number;
  feePercentage: number;
}

export interface MonthlyFinancials {
  month: string;
  monthDate: Date;
  grossRevenue: number;
  stripeFees: number;
  netRevenue: number;
  payouts: number;
}

export interface DailyPayout {
  date: string;
  dateObj: Date;
  amount: number;
  count: number;
  stripeAmount?: number;
  stripeCount?: number;
  pixAmount?: number;
  pixCount?: number;
}

export interface SubscriptionRecord {
  customerName: string;
  customerEmail: string;
  amount: number;
  date: Date;
  installmentNumber: number;
  invoiceId: string;
}

export interface FailedPayment {
  customerName: string;
  customerEmail: string;
  amount: number;
  attemptDate: Date;
  installmentNumber: number;
  invoiceId: string;
  failureReason?: string;
}

export class DataProcessor {
  calculateSubscriptionMRR(subscription: Stripe.Subscription): number {
    let totalMRR = 0;

    for (const item of subscription.items.data) {
      const price = item.price;
      const quantity = item.quantity || 1;

      if (!price || !price.recurring) {
        continue;
      }

      const unitAmount = (price.unit_amount || 0) / 100; // Convert from cents
      const interval = price.recurring.interval;
      const intervalCount = price.recurring.interval_count || 1;

      let monthlyAmount = 0;

      switch (interval) {
        case 'month':
          monthlyAmount = unitAmount / intervalCount;
          break;
        case 'year':
          monthlyAmount = unitAmount / (12 * intervalCount);
          break;
        case 'week':
          monthlyAmount = (unitAmount * 52) / (12 * intervalCount);
          break;
        case 'day':
          monthlyAmount = (unitAmount * 365) / (12 * intervalCount);
          break;
      }

      totalMRR += monthlyAmount * quantity;
    }

    return Math.round(totalMRR * 100) / 100;
  }

  calculateNewVsExistingMRR(subscriptions: Stripe.Subscription[]): MonthlyMRR[] {
    if (subscriptions.length === 0) return [];

    const subscriptionData = subscriptions
      .map((sub) => {
        const createdDate = new Date(sub.created * 1000);
        const canceledDate = sub.canceled_at ? new Date(sub.canceled_at * 1000) : null;
        const mrr = this.calculateSubscriptionMRR(sub);

        return {
          id: sub.id,
          createdDate,
          canceledDate,
          status: sub.status,
          mrr,
        };
      })
      .filter((sub) => sub.mrr > 0);

    // Collect all relevant months
    const monthsSet = new Set<string>();
    
    for (const sub of subscriptionData) {
      const createdMonth = startOfMonth(sub.createdDate);
      monthsSet.add(createdMonth.toISOString());

      if (sub.canceledDate) {
        const canceledMonth = startOfMonth(sub.canceledDate);
        let current = createdMonth;
        while (current <= canceledMonth) {
          monthsSet.add(current.toISOString());
          current = addMonths(current, 1);
        }
      } else {
        let current = createdMonth;
        const now = startOfMonth(new Date());
        while (current <= now) {
          monthsSet.add(current.toISOString());
          current = addMonths(current, 1);
        }
      }
    }

    const months = Array.from(monthsSet)
      .map((m) => new Date(m))
      .sort((a, b) => a.getTime() - b.getTime());

    const results: MonthlyMRR[] = [];

    for (const month of months) {
      const monthEnd = endOfMonth(month);
      let newMRR = 0;
      let existingMRR = 0;

      for (const sub of subscriptionData) {
        const isActiveInMonth =
          sub.createdDate <= monthEnd &&
          (!sub.canceledDate || sub.canceledDate >= month);

        if (isActiveInMonth) {
          const createdMonth = startOfMonth(sub.createdDate);
          if (createdMonth.getTime() === month.getTime()) {
            newMRR += sub.mrr;
          } else {
            existingMRR += sub.mrr;
          }
        }
      }

      results.push({
        month: format(month, 'MMM yyyy'),
        monthDate: month,
        newMRR: Math.round(newMRR * 100) / 100,
        existingMRR: Math.round(existingMRR * 100) / 100,
        totalMRR: Math.round((newMRR + existingMRR) * 100) / 100,
      });
    }

    return results;
  }

  calculateARR(subscriptions: Stripe.Subscription[]): number {
    let totalARR = 0;

    for (const sub of subscriptions) {
      if (sub.status === 'active' || sub.status === 'trialing') {
        const mrr = this.calculateSubscriptionMRR(sub);
        totalARR += mrr * 12;
      }
    }

    return Math.round(totalARR * 100) / 100;
  }

  calculateChurnMetrics(subscriptions: Stripe.Subscription[]): ChurnMetrics {
    const now = new Date();
    const startOfCurrentMonth = startOfMonth(now);

    let activeStartOfMonth = 0;
    let activeCurrent = 0;
    let churnedThisMonth = 0;

    for (const sub of subscriptions) {
      const createdDate = new Date(sub.created * 1000);
      const canceledDate = sub.canceled_at ? new Date(sub.canceled_at * 1000) : null;

      const wasActiveAtMonthStart =
        createdDate < startOfCurrentMonth &&
        (!canceledDate || canceledDate >= startOfCurrentMonth);

      const isActiveNow =
        createdDate < now && (!canceledDate || canceledDate >= now);

      if (wasActiveAtMonthStart) {
        activeStartOfMonth++;

        if (
          !isActiveNow &&
          canceledDate &&
          canceledDate >= startOfCurrentMonth &&
          canceledDate < now
        ) {
          churnedThisMonth++;
        }
      }

      if (isActiveNow) {
        activeCurrent++;
      }
    }

    const churnRate =
      activeStartOfMonth > 0
        ? (churnedThisMonth / activeStartOfMonth) * 100
        : 0;

    return {
      churnRate: Math.round(churnRate * 100) / 100,
      churnedCount: churnedThisMonth,
      activeCount: activeCurrent,
      previousActiveCount: activeStartOfMonth,
    };
  }

  calculateCustomerTrends(subscriptions: Stripe.Subscription[]): CustomerTrend[] {
    const monthlyCustomers = new Map<string, Set<string>>();

    for (const sub of subscriptions) {
      const createdDate = new Date(sub.created * 1000);
      const monthKey = format(startOfMonth(createdDate), 'yyyy-MM');
      
      if (!monthlyCustomers.has(monthKey)) {
        monthlyCustomers.set(monthKey, new Set());
      }
      
      monthlyCustomers.get(monthKey)!.add(sub.customer as string);
    }

    const sortedMonths = Array.from(monthlyCustomers.keys()).sort();
    let cumulative = 0;

    return sortedMonths.map((monthKey) => {
      const monthDate = new Date(monthKey + '-01');
      const newCustomers = monthlyCustomers.get(monthKey)!.size;
      cumulative += newCustomers;

      return {
        month: format(monthDate, 'MMM yyyy'),
        monthDate,
        newCustomers,
        cumulativeCustomers: cumulative,
      };
    });
  }

  calculateRevenueByPlan(
    subscriptions: Stripe.Subscription[],
    productNames: Record<string, string> = {}
  ): PlanRevenue[] {
    const planMRR = new Map<string, number>();

    for (const sub of subscriptions) {
      if (sub.status !== 'active' && sub.status !== 'trialing') {
        continue;
      }

      for (const item of sub.items.data) {
        const price = item.price;
        const product = price.product;
        
        let productName = 'Unknown Product';
        
        if (typeof product === 'string') {
          productName = productNames[product] || product;
        } else if (product && typeof product === 'object' && 'name' in product) {
          productName = product.name || 'Unknown Product';
        }

        const mrr = this.calculateSubscriptionMRR({
          ...sub,
          items: { data: [item], has_more: false, object: 'list', url: '' },
        });

        planMRR.set(productName, (planMRR.get(productName) || 0) + mrr);
      }
    }

    const totalMRR = Array.from(planMRR.values()).reduce((sum, mrr) => sum + mrr, 0);

    return Array.from(planMRR.entries())
      .map(([plan, mrr]) => ({
        plan,
        mrr: Math.round(mrr * 100) / 100,
        percentage: Math.round((mrr / totalMRR) * 10000) / 100,
      }))
      .sort((a, b) => b.mrr - a.mrr);
  }

  calculateFinancialMetrics(
    balanceTransactions: Stripe.BalanceTransaction[],
    payouts: Stripe.Payout[],
    balance: Stripe.Balance
  ): FinancialMetrics {
    let grossRevenue = 0;
    let stripeFees = 0;

    // Calculate from balance transactions
    for (const txn of balanceTransactions) {
      if (txn.type === 'charge' || txn.type === 'payment') {
        grossRevenue += txn.amount / 100; // Convert from cents
        stripeFees += txn.fee / 100;
      }
    }

    const netRevenue = grossRevenue - stripeFees;
    
    const totalPayouts = payouts
      .filter(p => p.status === 'paid')
      .reduce((sum, p) => sum + (p.amount / 100), 0);

    const availableBalance = balance.available
      .reduce((sum, b) => sum + b.amount, 0) / 100;

    const pendingBalance = balance.pending
      .reduce((sum, b) => sum + b.amount, 0) / 100;

    const feePercentage = grossRevenue > 0 ? (stripeFees / grossRevenue) * 100 : 0;

    return {
      grossRevenue: Math.round(grossRevenue * 100) / 100,
      stripeFees: Math.round(stripeFees * 100) / 100,
      netRevenue: Math.round(netRevenue * 100) / 100,
      totalPayouts: Math.round(totalPayouts * 100) / 100,
      pendingBalance: Math.round(pendingBalance * 100) / 100,
      availableBalance: Math.round(availableBalance * 100) / 100,
      feePercentage: Math.round(feePercentage * 100) / 100,
    };
  }

  calculateMonthlyFinancials(
    balanceTransactions: Stripe.BalanceTransaction[],
    payouts: Stripe.Payout[]
  ): MonthlyFinancials[] {
    const monthlyData = new Map<string, {
      gross: number;
      fees: number;
      payouts: number;
    }>();

    // Process balance transactions
    for (const txn of balanceTransactions) {
      if (txn.type === 'charge' || txn.type === 'payment') {
        const date = new Date(txn.created * 1000);
        const monthKey = format(startOfMonth(date), 'yyyy-MM');
        
        const existing = monthlyData.get(monthKey) || { gross: 0, fees: 0, payouts: 0 };
        existing.gross += txn.amount / 100;
        existing.fees += txn.fee / 100;
        monthlyData.set(monthKey, existing);
      }
    }

    // Process payouts
    for (const payout of payouts) {
      if (payout.status === 'paid' && payout.arrival_date) {
        const date = new Date(payout.arrival_date * 1000);
        const monthKey = format(startOfMonth(date), 'yyyy-MM');
        
        const existing = monthlyData.get(monthKey) || { gross: 0, fees: 0, payouts: 0 };
        existing.payouts += payout.amount / 100;
        monthlyData.set(monthKey, existing);
      }
    }

    // Convert to array and sort
    const results: MonthlyFinancials[] = [];
    for (const [monthKey, data] of Array.from(monthlyData.entries())) {
      const monthDate = new Date(monthKey + '-01');
      results.push({
        month: format(monthDate, 'MMM yyyy'),
        monthDate,
        grossRevenue: Math.round(data.gross * 100) / 100,
        stripeFees: Math.round(data.fees * 100) / 100,
        netRevenue: Math.round((data.gross - data.fees) * 100) / 100,
        payouts: Math.round(data.payouts * 100) / 100,
      });
    }

    return results.sort((a, b) => a.monthDate.getTime() - b.monthDate.getTime());
  }

  calculateDailyPayouts(payouts: Stripe.Payout[]): DailyPayout[] {
    const dailyData = new Map<
      string,
      { amount: number; count: number; stripeAmount: number; stripeCount: number }
    >();

    // Process payouts by arrival date
    for (const payout of payouts) {
      if (payout.status === 'paid' && payout.arrival_date) {
        const date = new Date(payout.arrival_date * 1000);
        const dateKey = format(date, 'yyyy-MM-dd');
        
        const existing =
          dailyData.get(dateKey) || { amount: 0, count: 0, stripeAmount: 0, stripeCount: 0 };
        const payoutAmount = payout.amount / 100;
        existing.amount += payoutAmount;
        existing.count += 1;
        existing.stripeAmount += payoutAmount;
        existing.stripeCount += 1;
        dailyData.set(dateKey, existing);
      }
    }

    // Convert to array and sort
    const results: DailyPayout[] = [];
    for (const [dateKey, data] of Array.from(dailyData.entries())) {
      const dateObj = new Date(dateKey);
      results.push({
        date: format(dateObj, 'dd/MM/yyyy'),
        dateObj,
        amount: Math.round(data.amount * 100) / 100,
        count: data.count,
        stripeAmount: Math.round(data.stripeAmount * 100) / 100,
        stripeCount: data.stripeCount,
        pixAmount: 0,
        pixCount: 0,
      });
    }

    return results.sort((a, b) => a.dateObj.getTime() - b.dateObj.getTime());
  }

  processSubscriptionRecords(invoices: Stripe.Invoice[]): SubscriptionRecord[] {
    const records: SubscriptionRecord[] = [];
    const customerPaymentCounts = new Map<string, number>();

    // Sort invoices by date to count installments correctly
    const sortedInvoices = [...invoices]
      .filter((inv) => inv.status === 'paid' && inv.subscription)
      .sort((a, b) => (a.created || 0) - (b.created || 0));

    for (const invoice of sortedInvoices) {
      const customer = invoice.customer as Stripe.Customer | string;
      const customerId = typeof customer === 'string' ? customer : customer.id;
      const customerName = typeof customer === 'object' ? (customer.name || 'N/A') : 'N/A';
      const customerEmail = typeof customer === 'object' ? (customer.email || 'N/A') : 'N/A';

      // Increment installment count for this customer
      const currentCount = customerPaymentCounts.get(customerId) || 0;
      customerPaymentCounts.set(customerId, currentCount + 1);

      records.push({
        customerName,
        customerEmail,
        amount: (invoice.amount_paid || 0) / 100,
        date: new Date((invoice.status_transitions?.paid_at || invoice.created) * 1000),
        installmentNumber: currentCount + 1,
        invoiceId: invoice.id,
      });
    }

    return records.sort((a, b) => b.date.getTime() - a.date.getTime());
  }

  processFailedPayments(invoices: Stripe.Invoice[]): FailedPayment[] {
    const failures: FailedPayment[] = [];
    const customerPaymentCounts = new Map<string, number>();

    // First, count all successful payments per customer to know installment numbers
    const allInvoicesSorted = [...invoices]
      .filter((inv) => inv.subscription)
      .sort((a, b) => (a.created || 0) - (b.created || 0));

    for (const invoice of allInvoicesSorted) {
      const customer = invoice.customer as Stripe.Customer | string;
      const customerId = typeof customer === 'string' ? customer : customer.id;

      if (invoice.status === 'paid') {
        const currentCount = customerPaymentCounts.get(customerId) || 0;
        customerPaymentCounts.set(customerId, currentCount + 1);
      }
    }

    // Now process failed invoices
    const failedInvoices = invoices.filter(
      (inv) =>
        inv.subscription &&
        (inv.status === 'open' || inv.status === 'uncollectible') &&
        inv.attempted &&
        !inv.paid
    );

    for (const invoice of failedInvoices) {
      const customer = invoice.customer as Stripe.Customer | string;
      const customerId = typeof customer === 'string' ? customer : customer.id;
      const customerName = typeof customer === 'object' ? (customer.name || 'N/A') : 'N/A';
      const customerEmail = typeof customer === 'object' ? (customer.email || 'N/A') : 'N/A';

      // Estimate installment number (count of paid + 1 for this failed attempt)
      const installmentNumber = (customerPaymentCounts.get(customerId) || 0) + 1;

      failures.push({
        customerName,
        customerEmail,
        amount: (invoice.amount_due || 0) / 100,
        attemptDate: new Date(
          (invoice.status_transitions?.finalized_at || invoice.created) * 1000
        ),
        installmentNumber,
        invoiceId: invoice.id,
        failureReason: invoice.last_finalization_error?.message || 'Pagamento falhou',
      });
    }

    return failures.sort((a, b) => b.attemptDate.getTime() - a.attemptDate.getTime());
  }
}

